### ISSUE TITLE
Increase QUERY_STRING max length

### ISSUE THREAD

**Issue by ur5us** [puma/puma#2273]
I am working an app that uses a GET request with a relatively long query string. Currently, Puma allows `1024 * 10` chars (presumably octets/bytes), s. https://github.com/puma/puma/blob/beab7f01b8c66524d2db68aeccbc34b31264b984/ext/puma_http11/puma_http11.c#L57. In our case that’s not enough. Now, if the above sounds familiar that’s probably due to a issue https://github.com/puma/puma/issues/2134 which was fixed by https://github.com/puma/puma/pull/2168. However, I don’t quite understand how this PR is actually fixing the original issue since the length of path is different from the length of the query params.

Ideally, I’d like to be able to configure the query string max length but that’s probably a lot of work since it’s currently determined at compile time if I understand correctly. Hence I’d like to propose simply increasing the max to 64k characters (or octets as per above). 32k or 16k would also work for us technically but 64k seems more future proof. 

I’ve tried to determine if there’s a max length defined in the RFCs but there doesn’t seem to be one, s.:

- https://tools.ietf.org/html/rfc7230
- https://tools.ietf.org/html/rfc3986#section-3.4

However, there’s some research mentioned in this [StackOverflow answer](https://stackoverflow.com/questions/812925/what-is-the-maximum-possible-length-of-a-query-string#812962). It suggests that that all modern browsers will submit query strings longer than the proposed 64k chars. 

In our case, the alternative to increasing the query string max length would be re-engineering a particular form. While this will work in the particular instance that lead to this feature request overall it rather feels like bandaid as we can’t control how our customers use our product and the lengthy forms they can build with it. In other words, this could happen again and then there’s probably not much that we could do technically.

Would you be interested in/open to a PR?

**Comment by dentarg** [comment]
> In our case, the alternative to increasing the query string max length would be re-engineering a particular form. While this will work in the particular instance that lead to this feature request overall it rather feels like bandaid as we can’t control how our customers use our product and the lengthy forms they can build with it. In other words, this could happen again and then there’s probably not much that we could do technically.

Are you able to share the reason for the forms needs to be GET and not POST forms? I'm genuinely curious :-)

**Comment by ur5us** [comment]
@dentarg Think of a filtering mechanism based on a rather large amount of filtering options that can be triggered individually. If you stick to REST then this is not a POST request.

**Comment by nateberkopec** [comment]
Maybe a C wizard could comment on a way we could allow variables like these to be set via an environment variable at compile time.

so basically:

```
PUMA_COMPILE_OPTS="querymaxlength=123123123" gem install puma
```

That seems like a good enough interface to solve this problem.

**Comment by ur5us** [comment]
@nateberkopec That would be useful indeed. Hopefully, some C wizard comes along indeed ;) In the interim, would you be open to increasing it to a fixed but larger value, e.g. 64k chars like I did here: https://github.com/ur5us/puma/commit/3ebdf54f1416809357783289e1a67561b5d158d2

Sidenote: is there a more complete writeup of how to setup, i.e. compile, link?, …, the specified version of Ragel? My Linux distribution of choice doesn’t provide it all. I’ve briefly looked at Ubuntu 20.04 but even that ‘only’ ships the stable 6.x branch. I’d be willing to go deeper and try figure it out if there’s a chance this change could get merged until that C wizard comes along.

**Comment by ur5us** [comment]
@nateberkopec I figured out all the Ragel/Colm bits and just went ahead to create a PR. Keen to hear your and potentially other people’s thoughts.

**Comment by catmando** [comment]
why can't this be a command line configuration?  Its not clear why the check needs to be via a constant value in C++.   The strings are allocated AFTER the check so what harm would there be in checking via some runtime value?

**Comment by maleksiuk** [comment]
If the maintainers want a compile time version, I have a commit that does it (https://github.com/maleksiuk/puma/commit/1c56e5f9ae9760b838d4ed8ce972f615a942e55b) and I could make a PR.

There are some pain points though:

1. It only works for CRuby.
1. If you forget to set the value on your first install, or want to change the value later, you have to go delete the gem folder and the extension folder before reinstalling. Not very user friendly.
1. There's no feedback that it understood and accepted the value, although I guess that can be added.

**Comment by nateberkopec** [comment]
> It only works for CRuby.

In the past I've merged stuff like this because if someone wants it in JRuby, they can come along and make a patch for it. We don't try to maintain superstrict parity in the extensions.

> If you forget to set the value on your first install, or want to change the value later, you have to go delete the gem folder and the extension folder before reinstalling. Not very user friendly.

I suppose this is true of any changes in compiled extensions, though.
 
> There's no feedback that it understood and accepted the value, although I guess that can be added.

That would be nice!

**Comment by nateberkopec** [comment]
If any C-hackers have feedback on @maleksiuk's approach I would love to hear it. From a general/non-C specific perspective this is definitely the way I would want to go to fix this.

**Comment by elct9620** [comment]
How about this way:

```c
#if defined(PUMA_QUERY_STRING_MAX_LENGTH)
DEF_MAX_LENGTH(QUERY_STRING, PUMA_QUERY_STRING_MAX_LENGTH);
#else
DEF_MAX_LENGTH(QUERY_STRING, (1024 * 10));
#end
```

When we install the ruby gem, I think it is possible to use:

```
gem install puma -- --with-cflags="-D PUMA_QUERY_STRING_MAX_LENGTH=64000"
```

**Comment by nateberkopec** [comment]
Works for me @elct9620, want to put in a PR?

**Comment by nateberkopec** [comment]
Or @maleksiuk

**Comment by elct9620** [comment]
I will create a PR later.

**Comment by dentarg** [comment]
Looks like this one was solved by https://github.com/puma/puma/pull/2485 (released with Puma v5.1.0) (see also https://github.com/puma/puma/pull/2514)

Docs at https://github.com/puma/puma/blob/master/docs/compile_options.md ([permalink](https://github.com/puma/puma/blob/3994981480a63ed070c68dcae581c029efbfcf9d/docs/compile_options.md))

**Comment by rafaelgurgel** [comment]
> How about this way:
> 
> ```c
> #if defined(PUMA_QUERY_STRING_MAX_LENGTH)
> DEF_MAX_LENGTH(QUERY_STRING, PUMA_QUERY_STRING_MAX_LENGTH);
> #else
> DEF_MAX_LENGTH(QUERY_STRING, (1024 * 10));
> #end
> ```
> 
> When we install the ruby gem, I think it is possible to use:
> 
> ```
> gem install puma -- --with-cflags="-D PUMA_QUERY_STRING_MAX_LENGTH=64000"
> ```

How would I go about having to cflags ? I need to set the query string and the request_uri. Already tried most of the possible string combinations with no success
